### PR TITLE
core: Implement parsing for Affine IR constructs

### PR DESCRIPTION
**Pull request by Groverkss**
The Parsing is complete even for semi-affine maps. Although, the AffineMap/AffineExpr IR constructs cannot handle them. The Affine IR constructs currently cannot handle some things like `(4 * 1) * d0`, because they cannot understand if `4 * 1` is a constant or not, but this will be improved in a separate PR with more tests.

### CODE REVIEW

File: xdsl/parser/affine_parser.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        self.resume_from_state(state)
+
+    def _parse_primary(self, dims: list[str], syms: list[str]) -> AffineExpr:
+        """
+        primary ::= `(` affine-expr `)`
+                  | bare-id
+                  | integer-literal
+                  | `-` primary
+        """
+        # Handle parenthesis
```
Comment (superlopuh) [inline xdsl/parser/affine_parser.py:27]:
```suggestion
        # Handle parentheses
```

File: xdsl/parser/core.py [hunk 2]
```diff
@@ -74,8 +75,11 @@
     ParametrizedAttribute,
     Data,
 )
+from xdsl.ir.affine import AffineMap
 from xdsl.utils.hints import isa
 
+from . import affine_parser as affine_parser
```
Comment (superlopuh) [inline xdsl/parser/core.py:81]: why do you need this? Does this not work?
```suggestion
from . import affine_parser
```

Comment (webmiche) [reply xdsl/parser/core.py:81]: Also can you make this an absolut import? We have been using these throughiut the repo, so I would stick to it.

File: xdsl/parser/affine_parser.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        """
+        primary ::= `(` affine-expr `)`
+                  | bare-id
+                  | integer-literal
+                  | `-` primary
+        """
+        # Handle parentheses
+        if self._parse_optional_token(Token.Kind.L_PAREN):
+            expr = self._parse_affine_expr(dims, syms)
+            self._consume_token(Token.Kind.R_PAREN)
```
Comment (math-fehr) [inline xdsl/parser/affine_parser.py:30]:
```suggestion
            self._parse_token(Token.Kind.R_PAREN)
```
`_consume_token` expects that you already know the kind of token you are parsing, and triggers an assertion.
`_parse_token` will parse the token, and raise an error if it is not the right kind.

**Review by webmiche — APPROVED** [review]
LGTM 👍

**Review by math-fehr — APPROVED** [review]
Nice!

**Pull request merged**